### PR TITLE
ARROW-17492: [C++] Hashing32/64 support for large var-binary types

### PR DESCRIPTION
**Pull request by rtpsw**
See https://issues.apache.org/jira/browse/ARROW-17492

### CODE REVIEW

**Comment by rtpsw** [comment]
cc @icexelloss

**Comment by rtpsw** [comment]
@michalursa, could you review this or refer to someone?

**Comment by rtpsw** [comment]
@westonpace, could you review this? You should be familiar with most of this code now.

**Comment by rtpsw** [comment]
> @westonpace, could you review this? You should be familiar with most of this code now.

Ping @westonpace

File: cpp/src/arrow/compute/exec/key_hash_test.cc [hunk 1]
```diff
@@ -35,9 +35,11 @@ namespace compute {
 
 class TestVectorHash {
  private:
-  static Result<std::shared_ptr<BinaryArray>> GenerateUniqueRandomBinary(
+  template <typename Type, typename ArrayType = typename TypeTraits<Type>::ArrayType>
+  static Result<std::shared_ptr<ArrayType>> GenerateUniqueRandomBinary(
```
Comment (westonpace) [inline cpp/src/arrow/compute/exec/key_hash_test.cc:39]: Can we add a quick comment (or a template constraint) that `Type` must be a binary type (e.g. Binary / LargeBinary)?  Maybe it should be obvious but I didn't quite catch it the first glance through.

**Review by westonpace — APPROVED** [review]
One super minor request.  Otherwise, thanks for this addition and thanks for making sure to add tests.

File: cpp/src/arrow/compute/exec/key_hash_test.cc (hunk 1 above)
Comment (rtpsw) [reply cpp/src/arrow/compute/exec/key_hash_test.cc]: Done.

File: cpp/src/arrow/compute/exec/key_hash_test.cc [hunk 2]
```diff
@@ -208,7 +219,8 @@ class TestVectorHash {
   }
 };
 
-TEST(VectorHash, Basic) {
+template <typename Type>
+void TestVectorHash() {
```
Comment (westonpace) [inline cpp/src/arrow/compute/exec/key_hash_test.cc:223]: Hmm, AppVeyor (Windows) is failing here:

```
C:/projects/arrow/cpp/src/arrow/compute/exec/key_hash_test.cc(239): error C2904: 'TestVectorHash': name already used for a template in the current scope
C:/projects/arrow/cpp/src/arrow/compute/exec/key_hash_test.cc(36): note: see declaration of 'arrow::compute::TestVectorHash'
[145/356] Building CXX object src\arrow\compute\kernels\CMakeFiles\arrow-compute-aggregate-test.dir\Unity\unity_0_cxx.cxx.obj
```

Maybe it doesn't like having both a method and a class named `TestVectorHash`?

```suggestion
void RunTestVectorHash() {
```

File: cpp/src/arrow/compute/exec/key_hash_test.cc [hunk 3]
```diff
@@ -219,12 +231,20 @@ TEST(VectorHash, Basic) {
...
 }
 
+TEST(VectorHash, BasicBinary) { TestVectorHash<BinaryType>(); }
+
+TEST(VectorHash, BasicLargeBinary) { TestVectorHash<LargeBinaryType>(); }
+
+TEST(VectorHash, BasicString) { TestVectorHash<StringType>(); }
+
+TEST(VectorHash, BasicLargeString) { TestVectorHash<LargeStringType>(); }
+
```
Comment (westonpace) [inline cpp/src/arrow/compute/exec/key_hash_test.cc:248]:
```suggestion
TEST(VectorHash, BasicBinary) { RunTestVectorHash<BinaryType>(); }

TEST(VectorHash, BasicLargeBinary) { RunTestVectorHash<LargeBinaryType>(); }

TEST(VectorHash, BasicString) { RunTestVectorHash<StringType>(); }

TEST(VectorHash, BasicLargeString) { RunTestVectorHash<LargeStringType>(); }

```

**Review by westonpace — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/compute/exec/key_hash_test.cc (hunk 2 above)
Comment (rtpsw) [reply cpp/src/arrow/compute/exec/key_hash_test.cc:223]: OK, let's try this.

**Review by westonpace — APPROVED** [review]
Thanks for adding this in.

**Pull request merged**